compiler/rustc_codegen_llvm/src/common.rs

```diff
@@ -392,3 +392,21 @@ pub(crate) fn get_dllimport<'tcx>(
     tcx.native_library(id)
         .and_then(|lib| lib.dll_imports.iter().find(|di| di.name.as_str() == name))
 }
+
+/// Extension trait for explicit casts to `*const c_char`.
+pub(crate) trait AsCCharPtr {
+    /// Equivalent to `self.as_ptr().cast()`, but only casts to `*const c_char`.
+    fn as_c_char_ptr(&self) -> *const c_char;
+}
+
+impl AsCCharPtr for str {
+    fn as_c_char_ptr(&self) -> *const c_char {
+        self.as_ptr().cast()
+    }
+}
```

Comment:
Isn't this wrong? `&str` can contain null bytes and casting it to c_char will lead to funny results in that case. It works for now because all used strings are well behaved and don not contain nulls.

Same with `[u8]`.

Reply:
in https://github.com/rust-lang/rust/pull/132319 it broken further: `c` literals (for which it's ok to take *c_char), you changed to &str, which don't have guarantees of cstr.

Reply:
And in #132319, all of the relevant FFI functions have been adjusted to take pointer/length pairs instead of nul-terminated strings, so there's no longer any need for those strings to be `c""` literals.

Reply:
Yes, but fragile. Can't find is it ok for StringRef to have nulls inside, if it later pass them as c strings, things will blow.